### ISSUE TITLE
ETag and ordering in tests

### ISSUE THREAD

**Issue by ajs6f** [trellis-ldp/trellis#237]
Reading [LdpBasicContainerTests](https://github.com/trellis-ldp/trellis/blob/master/components/test/src/main/java/org/trellisldp/test/LdpBasicContainerTests.java), I see many assertions where various ETags are compared, using `getFirstETag()`, `getSecondETag()`, etc. But in each test, only one method of the `setSecondETag(...)` group is being called. How are the other values being recorded? Is this a case in which the tests have to run in a certain order?

**Comment by acoburn** [comment]
Notwithstanding the naming of those methods, the order is unimportant (at least, it _should_ be unimportant -- I could be missing something). They are each ETags that are gathered/retrieved at different points of the test execution. The important thing being tested is that the particular ETag under evaluation is unique w/r/t the other ETags (whether they have been set or not).

**Comment by ajs6f** [comment]
I'm still a bit confused: when you write "at different points of the test execution" you mean "at different points within a single test"? Or "at different points as various tests are executed"?

**Comment by acoburn** [comment]
I mean "at different points as various tests are executed". So as test A runs, it will record its ETag and verify that it is different than the other ETags (all of which may be `null` if that test is run first). When test B runs (if the expectation is that the ETag changes), that ETag will be compared to the others, confirming that it is different. It shouldn't matter what order A and B are run in.

**Comment by ajs6f** [comment]
What if they run in parallel?

**Comment by ajs6f** [comment]
I should add some actual data here, I suppose 😁 . What I'm seeing is:
```
[ERROR] Failures:
[ERROR]   BasicContainmentIT.testCreateContainerViaPost Check that the ETag has been changed ==> expected: not equal but was: <W/"8c93f5c662d5fd86f28c57b0ce9beed3">
[ERROR]   BasicContainmentIT.testCreateContainerViaPut Check ETags 3 and 4 ==> expected: not equal but was: <W/"8c93f5c662d5fd86f28c57b0ce9beed3">
[ERROR]   BasicContainmentIT.testDeleteContainer Verify that the ETag value is different ==> expected: not equal but was: <W/"8c93f5c662d5fd86f28c57b0ce9beed3">
```
which is why I'm trying to figure out the "whens" of this thing.

**Comment by acoburn** [comment]
That points to a need to make the tests stronger.

**Comment by acoburn** [comment]
I think the `trellis-test` module could be significantly improved. It might take me a little bit of time, but I'll see what I can do with making the `LdpBasicContainerTests` completely independent. If the changes work for you, I can move on to the other tests.

**Comment by ajs6f** [comment]
I have now issued the above comment all of the positive emoji I could find.

**Comment by ajs6f** [comment]
Incidentally, am I correct in my reading of `trellis-http` that the only contribution of the backend to ETags for RDFSources is from `Resource::getModified`?

**Comment by acoburn** [comment]
It's a little more complicated. See [this method](https://github.com/trellis-ldp/trellis/blob/master/core/http/src/main/java/org/trellisldp/http/impl/RdfUtils.java#L107-L111) for the details.

There are so many complications with ETags and RDF that it makes the head swim. In this method, a hash gets built from the resource identifier + the modification date + a hash of the prefer header.

The inclusion of the prefer header is really important for any JavaScript application that makes any use of Prefer headers (i.e. fetch the resource triples and then, separately, fetch the member triples). Without that, the client gets a `304 Not Modified` with the second response, which would make it really hard to create any sort of half-way intelligent JS-based application.

IMO, ETag generation is an abyss, and the best you can do is "make it useful in the most contexts".

**Comment by ajs6f** [comment]
Actually, I think you just answered "yes" in a funny way-- my question was specifically to the contribution of the backend, and the backend knows nothing of `Prefer` and does not select identifiers. That's not to say that I'm not interested in the complete picture-- just at the moment I want to make absolutely positively super-100% no-doubt guaranteed sure that `trellis-cassandra` isn't making this harder than it needs to be. It sounds like correct behavior for a backend in this context is "make sure you got `Resource::getModified` completely correct", which is in itself an interesting question for `trellis-cassandra`.

That's because `t-c` is relying *entirely* on C* for the value of `getModified`. So I need to go now and check that that wasn't a really stupid thing to do. For example, what if C* quantizes that number on some weird basis? How even is that number calculated? I have no idea, actually, and I probably _should_. Okay, off to read a lot about C* time structures and recordation.

**Comment by acoburn** [comment]
Oh right. As for the contribution of the backend, it's just `Resource::getModified`. I would add that for all the `Resource` implementations I've written, I have that resource manage the modification date itself (i.e. if it's a file, don't believe the file's modification date; if it's an S3 blob, don't believe what AWS says is the modification date). I don't think it's necessarily wrong to rely on C* for that value, but it might result in more variability.

**Comment by ajs6f** [comment]
Right, it might be a terrible idea. I have been known to produce them with gusto and a touch of  the spice of fanaticism. 😆 

So why did I do it that way? Because I understood the semantics of `getModified` to be "this is when the last change was durable" and for something as async as the C* backend,  our code will never know that. Only C* itself knows. But are those the semantics of `getModified`? The Javadoc says only "Get the last modified date." which could mean that, or could mean "this is when the  last change that became durable was initiated", or a couple of other more obscure interpretations. I think you chose the "initiated" interpretation for the synchronous backends, eh?

**Comment by acoburn** [comment]
Yes, for the synchronous backends, I'm using the "initiated" interpretation. I think, however, that it is entirely worthwhile to pursue your approach. I really don't think your approach is wrong, and so the test suite shouldn't fail because of it.

**Comment by ajs6f** [comment]
Okay, cool. I'll keep it that way for now, but of course I will be researching C*'s timekeeping anyway. When this settles, we should enrich the Javadoc for `getModified` to explain the correct and incorrect interpretations.

**Comment by ajs6f** [comment]
Just for a further data point-- I switched to recording the timestamp myself (just the way you're doing it in the sync impls) and no change. Same failures for the same reason. I'll look more closely into `modified` values.

**Comment by acoburn** [comment]
Since the code in `trellis-219` branch has changed a bit, can you remind me which tests are failing? Is it still the create-via-POST, create-via-PUT and delete tests?

**Comment by acoburn** [comment]
Is there a branch of trellis-cassandra that I can test locally?

**Comment by ajs6f** [comment]
Yep, https://github.com/ajs6f/trellis-cassandra/tree/Containment.

**Comment by ajs6f** [comment]
Incidentally, I've stuffed that things to the gills with logging, as you would expect, and you should be able to see lines like:
```
2018-10-15 14:24:18,569 INFO  [stdout] (default task-44) DEBUG 18:24:18 Found modified = 2018-10-15T18:24:18.561Z for resource <trellis:data/dzymcfdacisujyzt-LdbBasicContainerTests/child4>
```
or 
```
2018-10-15 14:24:04,843 INFO  [stdout] (default task-55) DEBUG 18:24:04 Found modified = 2018-10-15T18:23:58.079Z for resource <trellis:data/dzymcfdacisujyzt-LdbBasicContainerTests>
```
to see that `trellis-cassandra` does seem to be retrieving sane modified-dates. The setup I'm using for C* allows me to inspect the db post-failure (I can detail that for you if you want) and the data in the backend are reasonable-- dates are distinct and have a sane order. They are only accurate to the millisecond because of limitations in the way I'm persisting them right now, but that's easily enough to distinguish them.

**Comment by ajs6f** [comment]
I should point out also that this thing won't even compile against Trellis `master` because of the changes we've made-- it needs to build against `trellis-219`.

**Comment by acoburn** [comment]
Yep, I'm building it w/ `trellis-219`

**Comment by acoburn** [comment]
I think I see what's going on. Your code stores the `modified` value in the `mutable` column, but when a containment triple is added to a parent container the `basiccontainment` column [is updated](https://github.com/ajs6f/trellis-cassandra/blob/Containment/impl/src/main/java/edu/si/trellis/cassandra/CassandraResourceService.java#L238-L240) (the `mutable` column isn't). So while the containment triple is definitely added to that parent resource, the modified date remains unchanged. In the triplestore impl, there is a special update query for HTTP creation and deletion operations that updates the modified value of the parent container. I would add that there's no expectation that the modified value of the parent exactly matches the modified value of the new resource. Does that help?

This logic gets even worse for Direct/Indirect containers, since there are additional lookups, but this should get you started for basic containment.

**Comment by ajs6f** [comment]
Goodness gracious. I don't know why I didn't get your eyes on this earlier. I completely misunderstood what you're doing with ETags-- I thought only the resource in question could contribute to its own ETag. I seem to recall dimly some kind of discussion around this with some of the LDP folks.

I'll fix this and see how much farther I get!

**Comment by ajs6f** [comment]
🔔  🛎  ch'ching. You just won this issue. A clean test. 

Do you want to merge the PR as a step towards "independentizing" the tests? I can't build against master because `LdpBasicContainerTests` is now so different.

**Comment by acoburn** [comment]
Excellent! I've merged #229 and the new artifacts will be in place shortly on the SNAPSHOT repository. Some similar modifications can be made to the other LDP test packages (Direct/Indirect containers), which I'll work on over the course of the week.
